Add upgrade form v4 to v5.1.0 (#1937)
emqtt/emq-docs-cn: en_US/deploy/upgrade-from-v4-to-v51.md

# Migrate from V4

> This section provides guidelines for migrating from EMQX 4.x to 5.1.0.

Reviewer: ```suggestion
This section provides guidelines for migrating from EMQX 4.4 to 5.1 or newer.
```

> EMQX 5.1.0 is not backward compatible with 4.x in management APIs and clustering APIs. However, most of the functionality is not significantly changed.

Reviewer: ```suggestion
EMQX 5 is not backward compatible with 4.x in management APIs and clustering APIs. However, most of the functionality is not significantly changed.
```

<!-- To learn about the new features we added to EMQX 5.1, check [New Features](../getting-started/new-features.md). -->

## Log

> Compared with 4.x, the most significant change in the log is the format. In 4.x, the log text is primarily plain, aiming for good human readability. However, starting from 5.1, we moved to structured logging without losing readability.

Reviewer: ```suggestion
Compared with 4.x, the most significant change in the log is the format. In 4.x, the log text is primarily plain, aiming for good human readability. Starting from 5, logging style is changed to be more structured, so it is more machine (log indexer).
```

> For instance, most log fields use underscores as word separators, making them more search-friendly. This format also helps log indexing tools to index the logs more effectively.

Reviewer: ```suggestion
The `msg` field value is a snake_case message, making them more search-friendly. This format also helps log indexing tools to index the logs more effectively, for instance:
```

```bash
2022-06-29T16:58:53.235042+02:00 [info] foo: bar, msg: msg_for_human_to_read_but_also_easy_to_index
```

Find more details in [Logs](../observability/log.md#log-examples).

## Default listeners

By default, the following listeners are no longer provided:

| Listeners            | Description                        |
| -------------------- | ---------------------------------- |
| MQTT-TCP 11883       | Listeners for backend applications |
| Management-HTTP 8081 | For REST API, merged to port 18083 |

## Plugins

> The previous official plugins have been migrated to EMQX as built-in functions. Custom plugins developed for version 4.x need to be adapted before they can be used in EMQX 5.1. The following is a comparison table between official plugins and existing functions:

Reviewer: ```suggestion
The previous official plugins have been migrated to EMQX as built-in functions. Custom plugins developed for version 4.x need to be adapted before they can be used in EMQX 5.1. The following is a comparison table between official plugins and built-in features.
```

| `GET /nodes`                     | `GET /nodes`                                | Partially compatible |                              |
| `GET /brokers`                   | -                                           | Incompatible         | merged to `GET /nodes`       |
| `GET /stats`                     | `GET /stats`                                | Partially compatible |                              |
| `GET /metrics`                   | `GET /metrics`                              | Partially compatible |                              |
| **Users/Alarms**                 |                                             |                      |                              |
| `GET /users`                     | `GET /users`                                | Partially compatible |                              |
| `GET /alarms{/activated}`        | `GET /alarms?activated={true,false}`        | Incompatible         |                              |
| `GET /alarms{/deactivated}`      | `GET /alarms?activated={true,false}`        | Incompatible         |                              |

> ## Auth/ACL

Reviewer: ```suggestion
## MQTT Client Authentication and Authorization
```

> **Function Entry**
> 
> The Auth and ACL plugins (emqx_auth_*) have been removed, and the corresponding functions are built into EMQX. Users can configure them through Dashboard or configuration files.

Reviewer: ```suggestion
```

**Change of concept**

> Auth is renamed to **Authentication**, and ACL is renamed **Authorization**. Now they are independent of each other.

Reviewer: ```suggestion
Auth is referred to as **Authentication**, and ACL is also referred to as **Authorization**.
```

**Data Migration**

We have kept the authentication methods and supported data sources from version 4.x, with only a few changes to the way they are used. For most Authenticator/Authorizer, you can continue to use the 4.x database when upgrading to version 5.1 without having to migrate existing data.

### Order of authentication/authorization chain

> When multiple authenticators or authorization checkers are enabled simultaneously, the checks are no longer performed according to the startup order but a fixed configuration order. The execution order can be adjusted in the configuration file and Dashboard.

Reviewer: ```suggestion
When multiple authenticators or authorization checkers are enabled simultaneously, the checks are no longer performed according to the startup order but a fixed order per configuration. The order can be adjusted from Dashboard.
```

### Variable Interpolation Syntax

Previously, the Auth plugins could use `%u` syntax to construct variable placeholders, dynamically interpolating client information into SQL statements, Redis query commands, and HTTP requests.
Now EMQX uses the new syntax `${}`, such as `${username}`, `${clientid}`, which is consistent with the rule SQL.

For supported placeholders, please refer to:

- [Authentication Placeholders](../access-control/authn/authn.md#authentication-placeholders)
- [Authorization Placeholders](../access-control/authz/authz.md#placeholders-in-data-queries)

Usage example:

```shell
# 4.x
# etc/emqx_auth_mysql.conf
auth.mysql.auth_query = select password from mqtt_user where username = '%u' limit 1

# 5.1
# emqx.conf
authentication = [
  {
    ...
    mechanism = "password_based"
    backend = "mysql"
    query = "SELECT password_hash, salt FROM mqtt_user where username = ${username} LIMIT 1"
  }
]
```

> ### Auth

Reviewer: ```suggestion
### MQTT Client Authentication
```

#### Remove the Anonymous Mechanism

> Remove the `allow_anonymous` configuration item. All client connections are allowed by default. If **add and enable** any authenticator, EMQX will perform an authentication check for all new connections.

Reviewer: ```suggestion
The `allow_anonymous` configuration is now deleted. All clients are allowed to connect by default. If **add and enable** any authenticator, EMQX will try to authenticate the clients.
```

> When a client matches the authentication data in all authenticators, the connection is rejected.

Reviewer: ```suggestion
After traversing the configured authentication chain, if none of the authenticator in the chain can decide that this client is allowed to connect, then the connection is rejected.
```

> Remove the `bypass_auth_plugins` configuration item. When a listener needs to skip authentication, it can be set through the `listeners.{type}.{name}.enable_authn = true | false` configuration item.

Reviewer: ```suggestion
The `bypass_auth_plugins` configuration is also deleted. When one wants to allow all clients to connect without authentication, they may set `listeners.{type}.{name}.enable_authn = false`.
```

#### Built-in database (Mnesia)

> 1. Mnesia is renamed to the built-in database;

Reviewer: ```suggestion
1. Mnesia is now referred to as the "built-in" database;
```

> 2. Only two search methods are available: username-based or clientid-based;

Reviewer: ```suggestion
2. Two search methods are available: username-based or clientid-based, but mixing is no longer supported.
```

> 3. The data format and REST API have changed. For more information, refer to `POST /authentication/{id}/users`.

Reviewer: ```suggestion
3. The REST APIs to manage the authentication data records are changed. For more information, refer to the API doc for `POST /authentication/{id}/users`.
```

#### HTTP

1. Use the JSON fields inside the response body instead of the HTTP response status codes to identify the authentication result;

> 2. Remove standalone super-user request. The super-user identity of the client is established through authentication in the response body.

Reviewer: ```suggestion
2. Removed standalone super-user request. The super-user identity of the client is established through authentication in the response body.
```

| 4.x | 5.1.0     | action              |
| --- | --------- | ------------------- |
| 1   | subscribe | subscribe           |
| 2   | publish   | publish             |
| 3   | all       | subscribe & publish |

If you want to continue using the data from in 4.x, please make necessary migrations manually.

**Data example in 5.1**

```bash
HSET mqtt_acl:emqx_u t/# subscribe
HSET mqtt_acl:emqx_u # all
HSET mqtt_acl:emqx_u a/1 publish
```

#### HTTP

1. Use the JSON fields in the response body instead of the HTTP response status codes to identify the authentication result.

**Success response status code:**

```shell
200 or 204
```

Other status codes or request failure will be treated as `ignore`.

**Success Response Body (JSON):**

| Name   | Type | Required | Description             |
| ------ | ---- | -------- | ----------------------- |
| result | Enum | true     | `allow | deny | ignore` |

```json
{
  "result": "deny"
}
```

## Rule Engine

Rule SQL is fully compatible with 4.x syntax, but the actions under the rule are split into built-in actions (republish, console) and data bridges (HTTP Server, MQTT Bridge).

> ## WebHook

Reviewer: ```suggestion
## HTTP Server
```

> The WebHook plugin (emqx_web_hook) has been removed. Use the HTTP Server data bridge in Data Integration instead.

Reviewer 2: HTTP Service？

> The WebHook plugin (emqx_web_hook) has been removed. Use the HTTP Server data bridge in Data Integration instead.

Reviewer: ```suggestion
The WebHook plugin (`emqx_web_hook`) is converted to a native feature, and it is now referred to as "HTTP Server" bridge.
```

## MQTT Bridge

The MQTT bridge plugin (emqx_bridge_mqtt) has been removed. Use the MQTT Sink and MQTT Source data bridge in Data Integration instead.

{% emqxee %}

## Offline messages

> The [offline messages](https://docs.emqx.com/en/enterprise/v4.4/rule/offline_msg_to_redis.html) provided in EMQX 4.x are based on an external database. EMQX plans to provide native offline messages (based on the built-in database) in future versions, so the offline messages for the external database have been removed from version 5.0.0.

Reviewer: ```suggestion
The [offline messages](https://docs.emqx.com/en/enterprise/v4.4/rule/offline_msg_to_redis.html) provided in EMQX 4.x are based on an external database. EMQX plans to provide native offline messages (based on the built-in database) in future versions, so the offline messages for the external database is no longer supported in version 5.
```

The upcoming native offline messaging feature will provide improved performance and reduce usage and maintenance costs. Stay tuned for more updates.

## Auto subscription (Server side subscriptions)

As of version 5.0.0, EMQX no longer provides [Auto subscription](https://docs.emqx.com/en/enterprise/v4.4/rule/get_subs_from_redis.html) (server side subscriptions) based on an external database.

{% endemqxee %}

## Prometheus

> Old Prometheus plug-in (emqx_prometheus) has been removed. In version 5.1, Prometheus scraping endpoint is enabled by default, and no authentication is required to scrap the metrics.

Reviewer: ```suggestion
The old plugin named `emqx_prometheus` has been converted to a native feature in version 5. Prometheus scraping endpoint is enabled by default, and no authentication is required to scrap the metrics.
```

| -                                  | erlang_vm_msacc_send_seconds_total              | New         |
| -                                  | erlang_vm_msacc_sleep_seconds_total             | New         |
| -                                  | erlang_vm_msacc_timers_seconds_total            | New         |
| -                                  | erlang_vm_statistics_dirty_cpu_run_queue_length | New         |
| -                                  | erlang_vm_statistics_dirty_io_run_queue_length  | New         |
| -                                  | erlang_vm_wordsize_bytes                        | New         |

## Gateway/Multi-Protocol Support

Clients of other protocols than MQTT (LwM2M, CoAP, STOMP, MQTT-SN) are no longer listed as MQTT clients on the dashboard **Connections** page and in the `GET /clients` API.
They can be found in **Extentions** -> **Gateways** section or listed with `GET /gateway/{name}/clients` API.

{% emqxce %}

## Telemetry

The telemetry plugin (emqx_telemetry) has been removed. Please configure it through the `telemetry {}` configuration item or in the Dashboard **System Settings** -> **Settings** page.

> {% endemqxce %}

Reviewer: in v5.0.26, emqx_telemetry is added back as an application, and there is not really much changed.
```suggestion
```
